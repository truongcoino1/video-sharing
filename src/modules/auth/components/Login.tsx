import React, { memo, useCallback, useMemo } from "react";
import { Button, useInput } from "@/modules/base";
import { useAuth } from "../hooks/useAuth";

const Login = () => {
  const { loading, login, error } = useAuth();
  const { value: email, handleOnChange: handleSetEmail } = useInput("");
  const { value: password, handleOnChange: handleSetPassword } = useInput("");

  const handleLogin = useCallback(
    async (event: React.FormEvent<HTMLFormElement>) => {
      event.preventDefault();
      login(email || "", password || "");
    },
    [email, password]
  );

  const isDisabled = useMemo(() => {
    return !email || !password;
  }, [email, password]);

  return (
    <div className="login-container">
      <form
        className="login-form flex-col flex lg:flex-row"
        onSubmit={handleLogin}
      >
        <input
          autoFocus
          name="email"
          type="email"
          placeholder="Email"
          aria-label="email"
          value={email}
          className="lg:mr-8 outline-none h-[40px] px-8 mt-12 lg:mt-0 border rounded-medium border-gray-300"
          onChange={handleSetEmail}
        />
        <input
          name="password"
          type="password"
          placeholder="Password"
          aria-label="password"
          value={password}
          className="lg:mr-8 h-[40px] outline-none px-8 border mt-12 lg:mt-0 rounded-medium border-gray-300"
          onChange={handleSetPassword}
        />
        <Button isLoading={loading} type="submit" isDisable={isDisabled} label="Login / Register" className="px-8 lg:mt-0 h-[40px] mt-12 rounded-medium" />
      </form>
    </div>
  );
};

export default memo(Login);
