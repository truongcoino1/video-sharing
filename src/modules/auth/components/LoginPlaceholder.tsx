import React, { memo } from "react";
import Skeleton from "react-loading-skeleton";

const LoginPlaceholder = (): JSX.Element => {
  return (
    <div className="flex login-placeholder">
      <Skeleton containerClassName="mr-12" className="w-[100px] min-w-[100px] lg:w-[200px] lg:min-w-[200px]  h-[40px] flex-shrink-0 overflow-hidden relative" />
      <Skeleton containerClassName="mr-12" className="w-[100px] min-w-[100px] lg:w-[200px] lg:min-w-[200px] mr-12 h-[40px] flex-shrink-0 overflow-hidden relative" />
      <Skeleton className="w-[80px] min-w-[80px] lg:w-[134px] lg:min-w-[150px] h-[40px] flex-shrink-0 overflow-hidden relative" />
    </div>
  );
};

export default memo(LoginPlaceholder);
