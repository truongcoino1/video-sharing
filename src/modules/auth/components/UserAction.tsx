import React, { memo } from 'react';
import Link from 'next/link';
import {useAuth} from "../hooks/useAuth"

const PATH_SHARE = '/share';

type UserActionProps = {
  email?: string;
};
const UserAction = ({email}: UserActionProps) => {
  const {logout} = useAuth();

  return (
    <div className="user-action-container">
      <span className="txt-welcome">Welcome {email}</span>
      <button
        className="btn-share-movie ml-12"
        disabled={window.location.pathname === PATH_SHARE}
      >
        {window.location.pathname === PATH_SHARE ? (
          'Share a movie'
        ) : (
          <Link href={PATH_SHARE}>Share a movie</Link>
        )}
      </button>
      <button className="btn-logout ml-12" onClick={logout}>
        Logout
      </button>
    </div>
  );
};

export default memo(UserAction);
