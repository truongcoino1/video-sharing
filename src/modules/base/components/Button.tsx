type ButtonProps = {
  isDisable?: boolean;
  label: string;
  className?: string;
  isLoading?: boolean;
  [x: string]: any;
};
export const Button = ({
  isDisable,
  label,
  className,
  isLoading,
  ...props
}: ButtonProps) => {
  return (
    <button
      className={`${className} flex items-center rounded-medium justify-center ${
        isDisable
          ? "opacity-50 cursor-not-allowed bg-dark-200 text-dark-high"
          : `relative before:z-0  text-dark-high hover:before:bg-light-100/20 bg-dark-200 before:h-full before:w-full before:absolute before:top-0 before:left-0 before:rounded-medium`
      } ${isLoading ? '!text-dark-200' : ''}`}
      disabled={isDisable}
      {...props}
    >
      {isLoading ? (
        <>
          <svg
            className="animate-spin w-20 h-20 text-light-100 absolute"
            viewBox="0 0 20 20"
            fill="none"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              fillRule="evenodd"
              clipRule="evenodd"
              d="M10 0C10.6497 0 11.1765 0.526724 11.1765 1.17647V4.11765C11.1765 4.76739 10.6497 5.29412 10 5.29412C9.35025 5.29412 8.82353 4.76739 8.82353 4.11765V1.17647C8.82353 0.526724 9.35025 0 10 0Z"
              fill="#ffffff"
            />
            <path
              fillRule="evenodd"
              clipRule="evenodd"
              d="M10 14.7059C10.6497 14.7059 11.1765 15.2326 11.1765 15.8824V18.8235C11.1765 19.4733 10.6497 20 10 20C9.35025 20 8.82353 19.4733 8.82353 18.8235V15.8824C8.82353 15.2326 9.35025 14.7059 10 14.7059Z"
              fill="#ffffff"
            />
            <path
              fillRule="evenodd"
              clipRule="evenodd"
              d="M2.93281 2.92497C3.39225 2.46553 4.13715 2.46553 4.59659 2.92497L6.67502 5.0034C7.13446 5.46284 7.13446 6.20774 6.67502 6.66718C6.21558 7.12662 5.47068 7.12662 5.01124 6.66718L2.93281 4.58875C2.47337 4.12931 2.47337 3.38441 2.93281 2.92497Z"
              fill="#ffffff"
            />
            <path
              fillRule="evenodd"
              clipRule="evenodd"
              d="M13.325 13.325C13.7844 12.8655 14.5293 12.8655 14.9887 13.325L17.0672 15.4034C17.5266 15.8628 17.5266 16.6077 17.0672 17.0672C16.6077 17.5266 15.8628 17.5266 15.4034 17.0672L13.325 14.9887C12.8655 14.5293 12.8655 13.7844 13.325 13.325Z"
              fill="#ffffff"
            />
            <path
              fillRule="evenodd"
              clipRule="evenodd"
              d="M0 10C0 9.35025 0.526724 8.82353 1.17647 8.82353H4.11765C4.76739 8.82353 5.29412 9.35025 5.29412 10C5.29412 10.6497 4.76739 11.1765 4.11765 11.1765H1.17647C0.526724 11.1765 0 10.6497 0 10Z"
              fill="#ffffff"
            />
            <path
              fillRule="evenodd"
              clipRule="evenodd"
              d="M14.7059 10C14.7059 9.35025 15.2326 8.82353 15.8824 8.82353H18.8235C19.4733 8.82353 20 9.35025 20 10C20 10.6497 19.4733 11.1765 18.8235 11.1765H15.8824C15.2326 11.1765 14.7059 10.6497 14.7059 10Z"
              fill="#ffffff"
            />
            <path
              fillRule="evenodd"
              clipRule="evenodd"
              d="M6.67502 13.325C7.13446 13.7844 7.13446 14.5293 6.67502 14.9887L4.59659 17.0672C4.13715 17.5266 3.39225 17.5266 2.93281 17.0672C2.47337 16.6077 2.47337 15.8628 2.93281 15.4034L5.01124 13.325C5.47068 12.8655 6.21558 12.8655 6.67502 13.325Z"
              fill="#ffffff"
            />
            <path
              fillRule="evenodd"
              clipRule="evenodd"
              d="M17.0672 2.92497C17.5266 3.38441 17.5266 4.12931 17.0672 4.58875L14.9887 6.66718C14.5293 7.12662 13.7844 7.12662 13.325 6.66718C12.8655 6.20774 12.8655 5.46284 13.325 5.0034L15.4034 2.92497C15.8628 2.46553 16.6077 2.46553 17.0672 2.92497Z"
              fill="#ffffff"
            />
          </svg>
          {label}
        </>
      ) : (
        label
      )}
    </button>
  );
};
