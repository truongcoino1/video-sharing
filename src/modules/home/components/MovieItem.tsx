import React, { useState, useCallback, memo } from "react";
import VideoPlaceholder from "./VideoPlaceholder";
import type { Movie } from "../types";

type MovieItemProps = {
  movie: Movie;
};
const MovieItem = ({ movie }: MovieItemProps): JSX.Element => {
  const [hasLoadIframe, setHasLoadIframe] = useState(false);

  const handleClickPlaceholder = useCallback(() => {
    setHasLoadIframe(true);
  }, []);

  return (
    <div className="movie-item lg:flex justify-between mb-12">
      {hasLoadIframe ? (
        <div className="lg:w-[350px] lg:h-[196px] rounded-large w-full flex-shrink-0 overflow-hidden relative">
          <iframe 
            className="w-full h-full rounded-large object-cover"
            title={movie.title}
            loading="lazy"
            allow="autoplay; fullscreen; accelerometer; encrypted-media; gyroscope; picture-in-picture"
            src={`https://www.youtube.com/embed/${movie.youtube_id}?autoplay=1`}
          />
        </div>
      ) : (
        <VideoPlaceholder
          thumbnail={movie.thumbnail}
          handleClick={handleClickPlaceholder}
        />
      )}
      <div className="lg:h-[196px] lg:ml-24 flex-grow lg:mt-0 lg:mb-0 mt-12 mb-24">
        <div className="font-bold text-[28px] overflow-hidden line-clamp-2 text-cyan-400">
          <a
            className="title-link"
            href={`https://www.youtube.com/watch?v=${movie.youtube_id}`}
            target="_blank"
            rel="noopener noreferrer"
          >
            {movie.title}
          </a>
        </div>
        <div className="flex items-center">
          <div className="shared-by text-light-low" title={movie.shared_by}>
            <span className="font-semibold text-16 text-light-high">Shared by:</span> {movie.shared_by}
          </div>
        </div>
        <div className="font-semibold text-16 text-light-high">Description:</div>
        <p className="text-14 overflow-hidden text-light-low line-clamp-2 video-description">{movie.description}</p>
      </div>
    </div>
  );
};

export default memo(MovieItem);
