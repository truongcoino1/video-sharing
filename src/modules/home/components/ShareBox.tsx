import React from "react";
import { useInput, Button } from "@/modules/base";
import { useMovies } from "../hooks/useMovies";
import { useAuthContext } from "@/modules/auth";
import SkeletonShareBox from "./SkeletonShareBox";

const ShareBox = () => {
  const { value: url, handleOnChange: handleSetUrl } = useInput("");
  const { shareMovie, isSharing } = useMovies();
  const { loading } = useAuthContext();

  if (loading) {
    return <SkeletonShareBox/>
  }

  return (
    <div className="sharebox-container mt-[100px] flex items-center justify-center">
      <form
        className="sharebox-form border relative border-gray-300 max-w-[calc(100vw_-_32px)] rounded-large w-[600px] p-24"
        onSubmit={(e) => {
          e.preventDefault();          
          shareMovie(url || "");
        }}
      >
        <span className="sharebox-title absolute top-[-20px] bg-light-100 font-medium text-24 px-8">
          Share a Youtube movie
        </span>
        <div className="sharebox-url flex items-center font-medium my-24">
          <label className="mr-12" htmlFor="url">
            Youtube URL:
          </label>
          <input
            autoFocus
            name="url"
            type="text"
            className="border border-gray-300 rounded-large flex-1 p-8"
            value={url}
            onChange={handleSetUrl}
          />
        </div>
        <div className="sharebox-submit flex items-center justify-center">
          <Button isLoading={isSharing} className="w-full p-8" label="Share" name="submit" type="submit" isDisable={!url} />
        </div>
      </form>
    </div>
  );
};

export default ShareBox;
