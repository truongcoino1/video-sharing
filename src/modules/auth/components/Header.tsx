/* eslint-disable @next/next/no-img-element */
"use client"
import React, { memo } from "react";
import Link from "next/link";
import Login from "./Login";
import UserAction from "./UserAction";
import { useAuthContext } from "../context";
import LoginPlaceholder from './LoginPlaceholder'

const PATH_HOME = "/";

const Header = () => {
  const { currentUser, loading } = useAuthContext();

  return (
    <header className="header-container p-[16px] lg:p-0 lg:h-[66px] lg:flex lg:justify-between lg:items-center border-b border-gray-300">
      <h1 className="title text-[36px] font-semibold mb-[8px] lg:mb-0">
        <Link className="no-style hover:text-black text-light-high flex" href={PATH_HOME}>
          <img src="/funny-movie.png" className="w-[48px] h-[48px] mr-12" alt="" />
          Funny Movies
        </Link>
      </h1>
      {loading ? <LoginPlaceholder/> : currentUser ? <UserAction email={currentUser.email} /> : <Login />}
    </header>
  );
};

export default memo(Header);
