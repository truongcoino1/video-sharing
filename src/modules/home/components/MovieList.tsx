import React, { memo, useEffect, useMemo } from "react";
import MovieItem from "./MovieItem";
import SkeletonMovie from "./SkeletonMovie";
import { useMovies } from "../hooks/useMovies";
import { Button } from "@/modules/base";
import { useAuthContext } from "@/modules/auth";
import { useSocket } from "@/modules/base/hooks/useSocket";

const fakeMovies = [1, 2, 3, 4, 5];

const MovieList = () => {
  const { movies, getMovies, period, setMovies } = useMovies();
  const { loading } = useAuthContext();
  const { isConnected, subscribe } = useSocket();

  useEffect(() => {
    getMovies({ refresh: true });
  }, []);

  useEffect(() => {
    if (isConnected) {
      subscribe("share-movie", (data: any) => {
        setMovies((prev) => [data, ...prev]);
      });
    }
  }, [isConnected]);

  const isShowLoadMore = useMemo(() => {
    return (
      !period.refresh &&
      !period.loadMore &&
      !period.limited &&
      movies.length > 0
    );
  }, [movies.length, period.limited, period.loadMore, period.refresh]);

  const isShowPlaceholder = useMemo(() => {
    return (period.refresh && movies.length === 0) || loading;
  }, [loading, movies.length, period.refresh]);

  return (
    <div className="movie-list-container pt-[32px]">
      {isShowPlaceholder
        ? fakeMovies.map((item) => <SkeletonMovie key={item} />)
        : movies.map((movie) => <MovieItem movie={movie} key={movie.id} />)}
      {isShowLoadMore && (
        <div className="btn-loadmore-wrapper flex items-center justify-center mt-24">
          <Button
            isLoading={period.loadMore}
            className="w-[150px] h-[40px] btn-loadmore"
            label="Load more"
            onClick={() => {
              getMovies({ loadMore: true })
            }}
          />
        </div>
      )}
      {period.limited && movies.length === 0 && (
        <div className="w-full flex justify-center items-center">
          No videos have been shared yet
        </div>
      )}
    </div>
  );
};

export default memo(MovieList);
