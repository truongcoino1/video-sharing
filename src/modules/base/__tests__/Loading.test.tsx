import { render, cleanup, fireEvent } from "@testing-library/react";
import { Loading } from "../components/Loading";

describe("Loading", () => {
  afterEach(cleanup);

  it("should render correct Loading", () => {
    const element = render(<Loading />);
    expect(element).toMatchSnapshot();
  });
});
