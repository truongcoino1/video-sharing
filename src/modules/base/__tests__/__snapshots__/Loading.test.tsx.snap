// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loading should render correct Loading 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="inline-block h-32 w-32 animate-spin rounded-full border-4 border-solid border-current border-r-transparent align-[-0.125em] motion-reduce:animate-[spin_1.5s_linear_infinite]"
        role="status"
      >
        <span
          class="!absolute !-m-px !h-px !w-px !overflow-hidden !whitespace-nowrap !border-0 !p-0 ![clip:rect(0,0,0,0)]"
        >
          Loading...
        </span>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="inline-block h-32 w-32 animate-spin rounded-full border-4 border-solid border-current border-r-transparent align-[-0.125em] motion-reduce:animate-[spin_1.5s_linear_infinite]"
      role="status"
    >
      <span
        class="!absolute !-m-px !h-px !w-px !overflow-hidden !whitespace-nowrap !border-0 !p-0 ![clip:rect(0,0,0,0)]"
      >
        Loading...
      </span>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
